compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala

```diff
@@ -71,189 +71,147 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
         .map(genJSClassDispatcher(classSym, _))
     }
 
-    def genConstructorExports(
-        classSym: Symbol): List[js.TopLevelMethodExportDef] = {
+    private sealed trait ExportKind
+
+    private object ExportKind {
+      case object Module extends ExportKind
+      case object JSClass extends ExportKind
+      case object Constructor extends ExportKind
+      case object Method extends ExportKind
+      case object Property extends ExportKind
+      case object Field extends ExportKind
+
+      def apply(sym: Symbol): ExportKind = {
+        if (isStaticModule(sym)) Module
+        else if (sym.isClass) JSClass
+        else if (sym.isConstructor) Constructor
+        else if (!sym.isMethod) Field
+        else if (jsInterop.isJSProperty(sym)) Property
+        else Method
+      }
+    }
 
-      val constructors = classSym.tpe.member(nme.CONSTRUCTOR).alternatives
+    private def checkSameKind(tups: Iterable[(jsInterop.ExportInfo, Symbol)]): Option[ExportKind] = {
+      assert(tups.nonEmpty, "must have at least one export")
 
-      // Generate exports from constructors and their annotations
-      val ctorExports = for {
-        ctor <- constructors
-        exp  <- jsInterop.registeredExportsOf(ctor)
-      } yield (exp, ctor)
+      val firstSym = tups.head._2
+      val overallKind = ExportKind(firstSym)
+      var bad = false
 
-      if (ctorExports.isEmpty) {
-        Nil
-      } else {
-        val exports = for {
-          (jsName, specs) <- ctorExports.groupBy(_._1.jsName) // group by exported name
-        } yield {
-          val ctors = specs.map(s => ExportedSymbol(s._2))
+      for ((info, sym) <- tups.tail) {
+        val kind = ExportKind(sym)
+
+        if (kind != overallKind) {
+          bad = true
+          reporter.error(info.pos, "export overload conflicts with export of " +
+              s"$firstSym: they are of different types ($kind / $overallKind)")
+        }
+      }
 
-          implicit val pos = ctors.head.pos
+      if (bad) None
+      else Some(overallKind)
+    }
 
-          val methodDef = withNewLocalNameScope {
-            genExportMethod(ctors, JSName.Literal(jsName), static = true)
-          }
+    private def checkSingleField(tups: Iterable[(jsInterop.ExportInfo, Symbol)]): Symbol = {
+      assert(tups.nonEmpty, "must have at least one export")
 
-          js.TopLevelMethodExportDef(methodDef)
-        }
+      val firstSym = tups.head._2
 
-        exports.toList
+      for ((info, _) <- tups.tail) {
+        reporter.error(info.pos, "export overload conflicts with export of " +
+            s"$firstSym: a field may not share its exported name with another export")
       }
+
+      firstSym
     }
 
-    def genJSClassExports(
-        classSym: Symbol): List[js.TopLevelJSClassExportDef] = {
-      for {
-        exp <- jsInterop.registeredExportsOf(classSym)
+    def genTopLevelExports(classSym: Symbol): List[js.TopLevelExportDef] = {
+      val exports = for {
```

Comment:
I suggest adding a `.toList` to the result of this for comprehension to address the issues with `.tail` above. If `exports` is a List, the `tups` should naturally become a `List` as well.